test/fakeintake/server/server.go

```diff
@@ -162,6 +176,25 @@ type postPayloadResponse struct {
 	Errors []string `json:"errors"`
 }
 
+func (fi *Server) cleanUpPayloads() {
+	for {
+		now := fi.clock.Now()
+		fi.mu.Lock()
+		for route, payloads := range fi.payloadStore {
+			n := 0
+			for _, payload := range payloads {
+				if now.Before(payload.Timestamp.Add(fi.retention)) {
+					fi.payloadStore[route][n] = payload
+					n++
+				}
+			}
+			fi.payloadStore[route] = fi.payloadStore[route][:n]
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
It looks like you are keeping oldest payloads with `[:n]`, cutting away last ones. What about using the following logic instead ?
```suggestion
            firstPayloadToKeepIndex := len(payloads) 
			for i, payload := range payloads {
				if now.Before(payload.Timestamp.Add(fi.retention)) {
					continue
				}
				// first time we find a young enough payload, we keep it and the followings
				firstPayloadToKeepIndex = i
				break
			}
			fi.payloadStore[route] = fi.payloadStore[route][firstPayloadToKeepIndex:]
```

Reply:
My function get rid of older payloads by moving payload to the beginning of the list every time it should be kept. The payload that have not been moved because too old are discarded. The order is preserved.

I think your function is an equivalent that can be slightly more efficient, but we need to be sure that payloads in the store are ordered, from the oldest to the newest. Which is probably the case here